### PR TITLE
tsuba: move parquet transformations into async tasks

### PR DESCRIPTION
**Pull request by tylershunt**
These transformations are single threaded and slow so try to overlap
them.

Some data:
For a graph with 1mil nodes, 100mil edges, 1 node property (string), and 5 edge properties (mix of types), the RDG is 2GB on disk.
 
Writes:
| Hosts     | Before |  After | Speedup |
| -------- | ------- |------| ----------|
| 1  |  41.54s | 22.40s   |  1.8X |
| 4  |   14.2s |  8.6s    | 1.6X |

Reads:
| Hosts     | Before |  After | Speedup |
| -------- | ------- |------| ----------|
| 1  |  58.4s | 23.6s   |  2.46X |
| 4  |   17.7s |  7.25s    | 2.45X |

(All machines have 32 cores, 25Gbps NICs, 256GB ram, numbers are average of 5 runs)

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

### CODE REVIEW

**Comment by tylershunt** [comment]
Updated on the read side as well. Will put in an enterprise PR that points to this to verify that it still passes all of our tests.

File: libtsuba/src/AsyncOpGroup.cpp
Comment (scober) [inline libtsuba/src/AsyncOpGroup.cpp:23]: Why `erase` instead of `pop_front()`?

File: libtsuba/src/RDGSlice.cpp
Comment (scober) [inline libtsuba/src/RDGSlice.cpp]: Are these comments meaningful? Temporary?

File: libtsuba/src/WriteGroup.cpp
Comment (scober) [inline libtsuba/src/WriteGroup.cpp:43]: I can't imagine it's a real issue, but won't the FIFO ordering of `AsyncOpGroup::FinishOne()` be suboptimal here?

**Review by scober — APPROVED** [review]
LGTM

File: libtsuba/src/AsyncOpGroup.cpp
Comment (tylershunt) [reply libtsuba/src/AsyncOpGroup.cpp:23]: because I already had an iterator so I figured why not 🤷 ?

File: libtsuba/src/WriteGroup.cpp
Comment (tylershunt) [reply libtsuba/src/WriteGroup.cpp:43]: I need to run the callbacks in order so that properties go into tables in the order that they did before. I was also hoping that  it's not a real issue, but you're right that we could relax this here to be `FinishAny`.

File: libtsuba/src/RDGSlice.cpp
Comment (tylershunt) [reply libtsuba/src/RDGSlice.cpp]: oh yeah, that's a note to self to make sliced reads async as well. I'll add that and remove this comment before it merges.

File: libtsuba/src/WriteGroup.cpp
Comment (scober) [reply libtsuba/src/WriteGroup.cpp:43]: I forgot about the callback ordering guarantee.

File: libtsuba/src/AsyncOpGroup.cpp
Comment (scober) [reply libtsuba/src/AsyncOpGroup.cpp:23]: Fair enough. This just reminded me of another piece of code you wrote recently (I don't remember where) that did a pop.

Comment (tylershunt) [reply libtsuba/src/AsyncOpGroup.cpp:23]: Bold of you to assume that I'm that consistent 😅

File: libtsuba/include/tsuba/AsyncOpGroup.h
Comment (witchel) [inline libtsuba/include/tsuba/AsyncOpGroup.h:34]: total bytes?

File: libtsuba/include/tsuba/ReadGroup.h
Comment (witchel) [inline libtsuba/include/tsuba/ReadGroup.h]: Never saw this trick before.  To avoid a dependence on arrow headers?

File: libtsuba/include/tsuba/ReadGroup.h
Comment (witchel) [inline libtsuba/include/tsuba/ReadGroup.h:25]: Is this equivalent to saying `on_complete` callbacks are called in FIFO order?

File: libtsuba/include/tsuba/ReadGroup.h
Comment (witchel) [inline libtsuba/include/tsuba/ReadGroup.h:38]: It is weird to see `std::make_shared` of a `shared_ptr`, but I guess `arrow::Table` is a abstract base class.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (witchel) [inline libtsuba/include/tsuba/ReadGroup.h:42]: We got an interview question right here!

Comment (witchel) [reply libtsuba/include/tsuba/ReadGroup.h:42]: Actually, I have a question.  I'd generally think to prefer `async `to `deferred `for IO.  Oh, but this operation only gets called on a return, so we defer it?

File: libtsuba/src/AsyncOpGroup.cpp
Comment (witchel) [inline libtsuba/src/AsyncOpGroup.cpp:32]: Is it important for this busy waiting to happen at full speed?  Might we consider the `_mm_pause();` intrinsic?  Or a `yield`?

File: libtsuba/src/AsyncOpGroup.cpp
Comment (witchel) [inline libtsuba/src/AsyncOpGroup.cpp:47]: Seeing dot-based structure initialization combined with std::move is very steampunk.

File: libtsuba/src/RDG.cpp
Comment (witchel) [inline libtsuba/src/RDG.cpp:431]: Do we still need this backward compatibility?

File: libtsuba/src/WriteGroup.cpp
Comment (witchel) [reply libtsuba/src/WriteGroup.cpp:43]: I wonder if we should use arrow Tables at all for properties.  Why not just use columns?

**Review by witchel — APPROVED** [review]
Its got the razzle, its got the dazzle, its got well-measured performance improvements.  Now that is what I call a PR.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h:42]: Yeah there's no real work in this function so it seemed fine to defer to call time when I was writing it;  probably it doesn't matter.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h:25]: Yeah; i'll change the comment just to say that.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h:38]: It's worse; here i'm making a pointer to a pointer as a piece of shared memory that my two lambdas below can use to communicate.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h]: yar.

File: libtsuba/src/AsyncOpGroup.cpp
Comment (tylershunt) [reply libtsuba/src/AsyncOpGroup.cpp:32]: `FinishOne` blocks until the next future it ready so this waiting isn't that busy.

File: libtsuba/src/AsyncOpGroup.cpp
Comment (tylershunt) [reply libtsuba/src/AsyncOpGroup.cpp:47]: ⚙️ 😎

File: libtsuba/src/WriteGroup.cpp
Comment (tylershunt) [reply libtsuba/src/WriteGroup.cpp:43]: We could talk about that; probably more churn than it's worth in this PR.

File: libtsuba/src/RDG.cpp
Comment (tylershunt) [reply libtsuba/src/RDG.cpp:431]: Not sure; would prefer not to change the actual functionality here though in case we do.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (witchel) [reply libtsuba/include/tsuba/ReadGroup.h:38]: Yeah, that does take you down the rabbit hole.  I'd at least add that as a comment.

File: libtsuba/include/tsuba/AsyncOpGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/AsyncOpGroup.h:34]: I'm working on a separate PR for stat tracking.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h:38]: I added a comment, and I actually templated the function as well (was my goal so that this works for return types besides `std::shared_ptr<arrow::Table>`).  I think the templating also makes the intent clearer since you don't see the confusing, duplicated `shared` on the line.

File: libtsuba/include/tsuba/ReadGroup.h
Comment (tylershunt) [reply libtsuba/include/tsuba/ReadGroup.h:42]: Doing some research I think `std::launch::async` will always create a new thread for the current version of g++ (no pooling at all seems to be the consensus), so a good practice is probably to try and used `std::launch::deferred` as much as possible to avoid thread creation overheads; especially for no-work tasks like this one.

Comment (witchel) [reply libtsuba/include/tsuba/ReadGroup.h:42]: Good to know.  A real shame they don't support thread pools.

**Pull request merged**
